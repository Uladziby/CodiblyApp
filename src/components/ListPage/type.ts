/** @format */

import { ITableItems } from "./../Table/type";

export interface IResponseProducts {
	page: number;
	per_page: number;
	total: number;
	total_pages: number;
	data: ITableItems[];
}

export interface IOptions {
	page: number;
	maxPage: number;
	isShowModal: boolean;
	choosedItem: ITableItems | undefined;
}

export type FormValues = {
	id: number;
};
