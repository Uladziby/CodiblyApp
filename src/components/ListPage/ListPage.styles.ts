/** @format */

import styled from "styled-components";

export const StyledContainer = styled.div`
	display: flex;
	justify-content: center;
	flex-direction: column;
	align-items: center;
	min-height: 200px;
	font-size: 1.5rem;
`;
