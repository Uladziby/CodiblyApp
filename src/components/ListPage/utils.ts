/** @format */

import { ITableItems } from "./../Table/type";
export const filterProducts = (id: number, products: ITableItems[]) =>
	id ? products.filter((item) => item.id == id) : products;
