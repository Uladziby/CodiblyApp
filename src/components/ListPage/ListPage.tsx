/** @format */

import { useEffect, useState } from "react";
import { useForm } from "react-hook-form";
import { useSearchParams } from "react-router-dom";
import { fetchProducts } from "../../api/api";
import { AInput } from "../AInput/AInput";
import { ModalContent } from "../ModalContent/ModalContent";
import { ModalWindow } from "../ModalWindow/ModalWindow";
import { Pagination } from "../Pagination/Pagination";
import { Table } from "../Table/Table";
import { ITableItems } from "../Table/type";
import { StyledContainer } from "./ListPage.styles";
import { FormValues, IOptions, IResponseProducts } from "./type";
import { filterProducts } from "./utils";

export const ListPage = () => {
	const [searchParams, setSearchParams] = useSearchParams();

	const [products, setProducts] = useState<ITableItems[]>();
	const [options, setOptions] = useState<IOptions>({
		page: Number(searchParams.get("page") || 1),
		maxPage: 1,
		isShowModal: false,
		choosedItem: { id: 0, year: 0, value: "", color: "", name: "" },
		error: "",
	});
	const { control, watch } = useForm<FormValues>();
	const { id } = watch();

	const onUpdateTable = () => {
		fetchProducts(options.page)
			.then((data: IResponseProducts) => {
				setProducts(data.data);
				setOptions((prevState) => ({
					...prevState,
					maxPage: data.total_pages,
				}));
			})
			.catch((err) => {
				setOptions((prevState) => ({
					...prevState,
					error: err,
				}));
			});
	};

	const onChangePageHandler = (page: number) => {
		setOptions((prevState) => ({
			...prevState,
			page,
		}));
	};

	const onShowModalHandler = (id: number) => {
		setOptions((prevState) => ({
			...prevState,
			isShowModal: !prevState.isShowModal,
			choosedItem: filterProducts(id, products!)[0],
		}));
	};

	const onCloseHandler = () => {
		setOptions((prevState) => ({
			...prevState,
			isShowModal: false,
		}));
	};

	useEffect(() => {
		setSearchParams(`page=${String(options.page)}`);
		onUpdateTable();
		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [options.page]);

	const visibleItems = filterProducts(id, products!);

	return (
		<>
			<StyledContainer data-testid="listPage">
				<form>
					<AInput type="number" name="id" placeholder="Type your id" control={control} />
				</form>
				{options.error ? (
					<StyledContainer>{options.error}</StyledContainer>
				) : (
					<Table onShowModalHandler={onShowModalHandler} items={visibleItems} />
				)}
				<Pagination
					onChangePageHandler={onChangePageHandler}
					currentPage={options.page}
					amountOfPages={options.maxPage}
				/>
			</StyledContainer>
			{options.isShowModal && (
				<ModalWindow onCloseHandler={onCloseHandler}>
					<ModalContent {...options.choosedItem!} />
				</ModalWindow>
			)}
		</>
	);
};
