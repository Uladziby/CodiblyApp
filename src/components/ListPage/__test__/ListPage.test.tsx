/** @format */

import { render, RenderResult } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import { ListPage } from "../ListPage";

describe("ListPage", () => {
	let renderResult: RenderResult;

	beforeEach(() => {
		renderResult = render(
			<MemoryRouter>
				<ListPage />
			</MemoryRouter>
		);
	});

	it("should render correctly", () => {
		const element = renderResult.getByTestId("listPage");

		expect(element).toBeInTheDocument();
	});

	it("snapshot", () => {
		const element = renderResult.getByTestId("listPage");

		expect(element).toMatchSnapshot();
	});
});
