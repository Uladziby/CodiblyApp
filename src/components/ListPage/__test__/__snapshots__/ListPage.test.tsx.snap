// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListPage snapshot 1`] = `
<div
  class="sc-hBxehG kJvEjx"
  data-testid="listPage"
>
  <form>
    <div
      class="sc-gswNZR gSPVJN"
    >
      <div
        class="sc-dkrFOg dYQiNc"
      >
        <input
          class="sc-bcXHqe bbDzJC"
          id="id"
          name="id"
          placeholder="Type your id"
          type="number"
          value=""
        />
      </div>
    </div>
  </form>
  <table
    class="sc-csuSiG cCQjHr"
  >
    <thead>
      <tr
        class="sc-eDWCr sc-bqWxrE fBUKkK hEvhnk"
      >
        <th
          class="sc-ksBlkl lhBUfD"
        >
          Id
        </th>
        <th
          class="sc-ksBlkl lhBUfD"
        >
          Name
        </th>
        <th
          class="sc-ksBlkl lhBUfD"
        >
          Year
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="sc-eDWCr hCJLul"
      >
        <td
          class="sc-ksBlkl lhBUfD"
        >
          List of product is an empty
        </td>
      </tr>
    </tbody>
  </table>
  <div
    class="sc-kDvujY essvcN"
  >
    <button
      class="sc-ipEyDJ ldiATk"
    >
      <svg>
        arrow_left.svg
      </svg>
    </button>
    <div>
      1
    </div>
    <button
      class="sc-ipEyDJ ldiATk"
    >
      <svg>
        arrow_right.svg
      </svg>
    </button>
  </div>
</div>
`;
